### PR TITLE
IA-2000 IA-1999 Add dettach

### PR DESCRIPTION
**Pull request by Qi77Qi**
https://broadworkbench.atlassian.net/browse/IA-2000
https://broadworkbench.atlassian.net/browse/IA-1999

1. add dettach disk
2. explicitly delete disk because once disk is detached, deleting runtime won't automatically delete disks anymore
3. Fix an edge case where when user interrupt `Stopping` with `Deleting`, we weren't doing deleting. Added https://broadworkbench.atlassian.net/browse/IA-2026 as well for addressing `Deleting` in the middle of `Starting`

---
Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; Delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

**Comment by Qi77Qi** [comment]
jenkins retest

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/Boot.scala [hunk 1]
```diff
@@ -73,6 +78,7 @@ object Boot extends IOApp {
       implicit val openTelemetry = appDependencies.openTelemetryMetrics
       implicit val dbRef = appDependencies.dbReference
 
+      val googleDependencies = appDependencies.googleDependencies
```
Comment (jdcanas) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/Boot.scala:81]: nice, was just thinking this class was getting ridiculously large

**Comment by Qi77Qi** [comment]
jenkins retest...fiab start failure

**Comment by Qi77Qi** [comment]
fiab start failure..jenkins retest

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/GceRuntimeMonitor.scala [hunk 2]
```diff
@@ -55,52 +63,90 @@ class GceRuntimeMonitor[F[_]: Parallel](
...
+      haltWhenTrue = (Stream
+        .eval(clusterQuery.getClusterStatus(runtimeAndRuntimeConfig.runtime.id).transaction.map { newStatus =>
+          if (action == RuntimeStatus.Deleting) // Deleting is not interruptible
+            false
+          else {
+            newStatus != Some(action) && newStatus == Some(
+              RuntimeStatus.Deleting
+            ) // Interrupt Stopping if new Deleting request comes in
+          }
+        }) ++ Stream.sleep_(5 seconds)).repeat
```
Comment (jdcanas) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/GceRuntimeMonitor.scala:75]: this should probably be in a config

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/GceRuntimeMonitor.scala [hunk 3]
```diff
@@ -55,52 +63,90 @@ class GceRuntimeMonitor[F[_]: Parallel](
...
+      _ <- if (pollCheckSupportedStatuses.contains(action))
+        F.unit
+      else F.raiseError(new Exception(s"Monitoring ${action} with pollOperation is not supported"))
```
Comment (jdcanas) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/GceRuntimeMonitor.scala:78]: this sounds like it should be enforced with the data type passed in as opposed to a runtime validation

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/GceRuntimeMonitor.scala (hunk 2 above)
Comment (Qi77Qi) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/GceRuntimeMonitor.scala:75]: there are way too many intervals like this...I think I added these in configs in early PRs..but I'm leaning toward having them in code now

**Review by Qi77Qi — COMMENTED** [review]
I think this is fine...I'm hoping slowly we can migrate the other transitions to `pollCheck` as well..at that point, having extra data types here will make things more complicated

File: http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/dao/google/MockGoogleDataprocDAO.scala [hunk 4]
```diff
@@ -137,23 +137,21 @@ class MockGoogleDataprocDAO(ok: Boolean = true) extends GoogleDataprocDAO {
...
-      val existingMasterInstance = instances.get(clusterName).flatMap(_.get(Master))
-
-      instances += (clusterName -> mutable.Map(Master -> existingMasterInstance.getOrElse(Set.empty),
-                                               Worker -> existingWorkerInstances.getOrElse(Set.empty),
-                                               SecondaryWorker -> secondaryWorkerInstances))
-    }
+    val secondaryWorkerInstances = numPreemptibles
+      .map(num =>
+        List
+          .tabulate(num) { i =>
```
Comment (jdcanas) [inline http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/dao/google/MockGoogleDataprocDAO.scala:143]: TIL

**Review by jdcanas — APPROVED** [review]

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/GceRuntimeMonitor.scala [hunk 5]
```diff
@@ -55,52 +63,90 @@ class GceRuntimeMonitor[F[_]: Parallel](
...
+      _ <- computePollOperation
+        .pollOperation(googleProject,
+                       operation,
+                       config.pollingInterval,
+                       config.pollCheckMaxAttempts,
+                       Some(haltWhenTrue))(
+          handlePollCheckCompletion(monitorContext, runtimeAndRuntimeConfig),
+          handlePollCheckTimeout(monitorContext, runtimeAndRuntimeConfig),
+          handlePollCheckWhenInterrupted(monitorContext, runtimeAndRuntimeConfig)
```
Comment (jdcanas) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/GceRuntimeMonitor.scala:87]: hmm... this monitoring functionality is getting quite complex... we're basically implementing our own event system. There does seem to be any good alternatives though

**Comment by jdcanas** [comment]
jenkins retest

**Review by rtitle — APPROVED** [review]

**Pull request merged**